activesupport/lib/active_support/cache/serializer_adapter.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module Cache
+    class SerializerAdapter # :nodoc:
+      def initialize(serializer:, compressor:, adapt_dump: true)
+        @serializer = serializer
+        @compressor = compressor
+        @adapt_dump = adapt_dump
+        @serializer_dumps_compressed = @serializer.respond_to?(:dump_compressed)
+      end
+
+      def dump(entry)
+        if !@adapt_dump
+          return @serializer.dump(entry)
+        end
+
+        dump_compressed(entry, Float::INFINITY)
+      end
+
+      def dump_compressed(entry, threshold)
+        if !@adapt_dump
+          return @serializer_dumps_compressed ? @serializer.dump_compressed(entry, threshold) : @serializer.dump(entry)
+        end
+
+        value = entry.value
+
+        if value.instance_of?(String) && type = STRING_ENCODINGS.key(value.encoding)
+          payload = value
+        else
+          type = OBJECT_DUMP_TYPE
+          payload = @serializer.dump(Cache::Entry.new(value))
+        end
+
+        if payload.bytesize >= threshold
+          compressed = @compressor.deflate(payload)
+          if compressed.bytesize < payload.bytesize
+            type = -type
+            payload = compressed
+          end
+        end
+
+        version = entry.version
+
+        if version && (version.empty? || version.encoding != Encoding::UTF_8)
+          version = Marshal.dump(version)
+          version_length = -version.bytesize
+        else
+          version_length = version&.bytesize || 0
+        end
+
+        packed = [SIGNATURE, type, entry.expires_at || -1.0, version_length].pack(PACKED_TEMPLATE)
+        packed << version if version
+        packed << payload
+      end
+
+      def load(dumped)
+        return @serializer.load(dumped) if !dumped?(dumped)
+
+        type = dumped.unpack1(PACKED_TYPE_TEMPLATE)
+        expires_at = dumped.unpack1(PACKED_EXPIRES_AT_TEMPLATE)
+        version_length = dumped.unpack1(PACKED_VERSION_LENGTH_TEMPLATE)
+
+        expires_at = nil if expires_at < 0
+
+        if version_length > 0
+          version = dumped.byteslice(PACKED_VERSION_INDEX, version_length).force_encoding(Encoding::UTF_8)
+        elsif version_length < 0
+          version_length = version_length.abs
+          version = Marshal.load(dumped.byteslice(PACKED_VERSION_INDEX, version_length))
+        end
+
+        payload = dumped.byteslice((PACKED_VERSION_INDEX + version_length)..)
+
+        value =
+          case type
+          when OBJECT_DUMP_TYPE
+            -> { @serializer.load(payload).value }
+          when COMPRESSED_OBJECT_DUMP_TYPE
+            -> { @serializer.load(@compressor.inflate(payload)).value }
+          when STRING_DUMP_TYPES
+            payload.force_encoding(STRING_ENCODINGS[type])
+          when COMPRESSED_STRING_DUMP_TYPES
+            -> { @compressor.inflate(payload).force_encoding(STRING_ENCODINGS[type.abs]) }
+          else
+            return nil
+          end
```

Comment:
I don't have very strong feelings one way or another, but should we use this case statement as type of factory?  It feels like we could eliminate the above conditional in the `value` method.  Additionally this requires us to allocate an extra proc object when we could just select a subclass of `Cache::Entry` that has a conveniently implemented `value` method.

Reply:
I agree here. Only the 6.1 format is stuck with the current implementation of `Entry`. We could use another class.

Reply:
I've added a commit (https://github.com/rails/rails/pull/48451/commits/4291e606fe6915647259d4047cad7e9fb6f4abbb) that defines specialized `Entry` subclasses.  But, in my opinion, the code is less clear.  Deserialization logic is strongly coupled with serialization logic.  Moving deserialization logic out of `SerializerAdapter` and away from `SerializerAdapter#dump` obscures this relationship.

If we were not implementing lazy deserialization, I think we would simply write the `case` statement as it was previously minus the `-> { }` wrappers.  To me, using `-> { }` to implement lazy evaluation (basically, a [thunk](https://en.wikipedia.org/wiki/Thunk)) makes sense.

Reply:
Using a thunk is fine, but the concrete classes give us a chance to add some features specific to the strategy without leaking in to the scope of the proc (for example the `@processed` IV you added in the subclasses).  I think we could do the same tricks with [this conditional](https://github.com/rails/rails/commit/4291e606fe6915647259d4047cad7e9fb6f4abbb#diff-545287677174ddf9aff59851be5b659a050f2ff5db316c0a9a57dcd24b72dcd0L1051) because we [should know at allocation time whether or not the entry is compressed](https://github.com/rails/rails/commit/4291e606fe6915647259d4047cad7e9fb6f4abbb#diff-545287677174ddf9aff59851be5b659a050f2ff5db316c0a9a57dcd24b72dcd0L1046) (additionally we could add the `@processed` treatment in the compressed cases)

Reply:
> without leaking in to the scope of the proc

By this, do you mean holding on to references that were in scope when the proc was created?  The `Entry` instance should be short-lived, just long enough to check `Entry#expired?` and `Entry#mismatched?` such as [in `Cache#read`](https://github.com/rails/rails/blob/d9f22a40e2bf3778b47a2d9fbb065bb7b10257d4/activesupport/lib/active_support/cache.rb#L431-L441).  But, if it is a concern, I can minimize the scope by isolating the `case` statement in its own method.

> I think we could do the same tricks with [this conditional](https://github.com/rails/rails/commit/4291e606fe6915647259d4047cad7e9fb6f4abbb#diff-545287677174ddf9aff59851be5b659a050f2ff5db316c0a9a57dcd24b72dcd0L1051) because we [should know at allocation time whether or not the entry is compressed](https://github.com/rails/rails/commit/4291e606fe6915647259d4047cad7e9fb6f4abbb#diff-545287677174ddf9aff59851be5b659a050f2ff5db316c0a9a57dcd24b72dcd0L1046)

That conditional and the `@compressed` ivar are used by the 6.1 cache format, and must remain until we drop support for the 6.1 format.

The 7.0 format and the code in this PR does not make use of them.  If it is more clear, an alternative way to write the method is:

```ruby
def value
  if @value.instance_of?(Proc)
    @value = @value.call
  elsif compressed?
    uncompress(@value)
  else
    @value
  end
end
```

> (additionally we could add the `@processed` treatment in the compressed cases)

In the previous commit that I pushed (https://github.com/rails/rails/commit/4291e606fe6915647259d4047cad7e9fb6f4abbb), `@processed` does apply to the compressed cases ([here](https://github.com/rails/rails/commit/4291e606fe6915647259d4047cad7e9fb6f4abbb#diff-545287677174ddf9aff59851be5b659a050f2ff5db316c0a9a57dcd24b72dcd0R1153-R1154) and [here](https://github.com/rails/rails/commit/4291e606fe6915647259d4047cad7e9fb6f4abbb#diff-545287677174ddf9aff59851be5b659a050f2ff5db316c0a9a57dcd24b72dcd0R1171-R1172)).